File: Matrix sensing/Python/LRMS_demo.py
import numpy as np
import time
from generate_matrix import generate_matrix
from GNMR_sensing import GNMR_sensing,\
    INIT_WITH_ATb, INIT_WITH_RANDOM, INIT_WITH_USER_DEFINED

def run_demo():
    # experiment configurations
    n1 = 50
    n2 = 60
    rank = 3
    condition_number = 1e1
    oversampling_ratio = 1.5
    singular_values = np.linspace(1, condition_number, rank)
    print("n1, n2:", n1, n2)
    print("oversampling ratio:", oversampling_ratio)
    print("singular values:", singular_values)

    # algorithm options (for documentation and more options, see GNMR_completion.py)
    options = {
        # general
        'verbose' : True,
        'alpha' : 1,  # (1: setting variant, 0: averaging variant, -1: updating variant)
        # number of iterations
        'max_outer_iter' : 100, 
        'max_inner_iter': 2000,
        # early stopping criteria (-1 to disable a criterion)
        'stop_relRes':  5e-14,
        'stop_relDiff': 5e-14,
    }

    # calculate full matrix, measurement operator and corresponding observed vector
    p = oversampling_ratio * rank * (n1 + n2 - rank) / (n1 * n2)
    X0 = generate_matrix(n1, n2, singular_values)
    m = int(np.ceil(oversampling_ratio * rank * (n1+n2-rank)))
    A = np.random.normal(0, 1.0/np.sqrt(m), (m, n1*n2))
    b = A @ np.ravel(X0)

    # run GNMR
    start = time.time()
    X_hat, iter, _, _ = GNMR_sensing(b, A, n1, n2, rank, **options)
    end = time.time()

    # report
    print("iter: ", iter, ". elapsed time:", end - start)
    true_error = np.linalg.norm(X_hat - X0, ord='fro') / np.linalg.norm(X0, ord='fro')
    observed_error = np.linalg.norm((A @ np.ravel(X_hat) - b)) / np.linalg.norm(b)
    print('true error: {}, observed error: {}'.format(true_error, observed_error))


if __name__ == '__main__':
    run_demo()

File: Matrix sensing/MATLAB/LRMS_demo.m
%% configurations

% experiment configurations
n1 = 50;
n2 = 60;
r = 3;
condition_number = 1e1;  % condition number
oversampling_ratio = 1.5;
singular_values = linspace(1, condition_number, r);

% algorithm options (for more options, see GNMR_sensing.m)
clear opts
opts.verbose = 1;               % display intermediate results
opts.alpha = 1;                 % variant parameter (e.g., 1: setting, 0: averaging, -1: updating)
% number of iterations
opts.max_outer_iter = 100;      % maximal number of outer iterations
opts.max_inner_iter = 2000;     % maximal number of inner iterations for the LSQR solver
% stopping criteria (-1 to disable a criterion)
opts.stop_relRes = 1e-14;   	% small relRes threshold
                                % (relRes = ||X_hat - X||_F/||X_hat||_F on the observed entires)
opts.stop_relDiff = 1e-14;      % small relative X_hat difference threshold


%% run experiment
format long;
fprintf('\n n1,n2: %4d,%4d. rank: %2d. condition number: %e \n oversampling ratio: %e\n\n', ...
    n1, n2, r, condition_number, oversampling_ratio);

rng_value = 2021;
rng('default');
rng(rng_value);

% generate low rank matrix X0
[X0, U0, V0] = generate_matrix(n1,n2,singular_values);

% generate sensing operator of Gaussian measurements
m = ceil(oversampling_ratio * r * (n2+n1-r));  % number of observations
A = normrnd(0, 1.0/sqrt(m), m, n1*n2);

% compute b, the observed linear measurements, according to A
X0_vec = X0(:);
b = A * X0_vec;

% run GNMR
[X_hat, ~, iter] = GNMR_sensing(b, A, n1, n2, r, opts);

% report
true_error = norm(X_hat - X0, 'fro') / norm(X0, 'fro');
fprintf('true error %8d, iter %3d\n', true_error, iter);

